### PR TITLE
open tunnels on demand

### PR DESCRIPTION
**Pull request by kopiczko**
Goal here is to make things less mutable. I.e. Forwarder is immutable. Tunnel is created on demand by ForwardPort. The Forwarder can be reused.

### CODE REVIEW

File: tunnel.go [hunk 1]
```diff
@@ -2,98 +2,78 @@ package k8sportforward
...
 }
 
-// Close disconnects a tunnel connection.
-func (t *Tunnel) Close() {
-	close(t.stopChan)
-}
-
 // ForwardPort opens a tunnel to a kubernetes pod.
-func (t *Tunnel) ForwardPort() error {
+func (f *Forwarder) ForwardPort(config TunnelConfig) (*Tunnel, error) {
```
Comment (kopiczko) [inline tunnel.go:47]: So now `ForwardPort` creates `Tunnel`s.

File: tunnel.go [hunk 2]
```diff
@@ -103,12 +83,31 @@ func (t *Tunnel) ForwardPort() error {
...
+	Local int
+
+	stopChan chan struct{}
+}
+
+// Close disconnects a tunnel connection.
+func (t *Tunnel) Close() error {
+	close(t.stopChan)
+	return nil
+}
```
Comment (kopiczko) [inline tunnel.go]: Tunnel is self-contained and can be closed. Forwarder can be reused.

**Review by xh3b4sd — APPROVED** [review]
Fine with me.

File: tunnel.go [hunk 3]
```diff
@@ -103,12 +83,31 @@ func (t *Tunnel) ForwardPort() error {
...
+
+type Tunnel struct {
+	TunnelConfig
+	Local int
+
+	stopChan chan struct{}
+}
+
+// Close disconnects a tunnel connection.
+func (t *Tunnel) Close() error {
```
Comment (fgimenez) [inline tunnel.go]: Can this return something != nil?

**Review by fgimenez — COMMENTED** [review]
Changes LGTM, small nit. We should also try it from chart-operator before merging, there were issues with adding k8s deps when trying to reuse upstream code, see https://github.com/giantswarm/chart-operator/pull/34#issuecomment-376923461

**Comment by kopiczko** [comment]
Before merging I'd like to know @fgimenez opinion on the matter.
I would also like to know what should I update after merging.

File: tunnel.go (hunk 3 above)
Comment (kopiczko) [reply tunnel.go]: No, but it fulfils io.Closer interface now.

Comment (kopiczko) [reply tunnel.go]: See https://golang.org/pkg/io/#Closer Close return an error. I think it's always good to conform to basic interfaces. WDYT?

**Comment by kopiczko** [comment]
@fgimenez re dependencies I don't see potential problem here, chart-operator actually imports superset of packages listed in this PR:

https://github.com/giantswarm/chart-operator/blob/74ab94859550ef2b839d57ad73a3e9df5a4ebc47/Gopkg.lock#L738-L821

https://github.com/giantswarm/k8sportforward/pull/4/files#diff-bd247e83efc3c45ae9e8c47233249f18R257

~~`https://github.com/giantswarm/k8sportforward/pull/4/files#diff-bd247e83efc3c45ae9e8c47233249f18R178`~~

**Comment by fgimenez** [comment]
The problem we had was related to k8s.io/api libraries, not client-go, and this PR adds some more  https://github.com/giantswarm/k8sportforward/pull/4/files#diff-bd247e83efc3c45ae9e8c47233249f18R177. I think we should test it before merging, maybe we could create branches on both chart-operator and helmclient to test the right deps, WDYT?

File: tunnel.go (hunk 3 above)
Comment (fgimenez) [reply tunnel.go]: Ok sounds good :+1: maybe a comment would make things more clear.

**Comment by kopiczko** [comment]
OK, let's check that

File: tunnel.go [hunk 4]
```diff
@@ -15,85 +14,65 @@ import (
...
-		return nil, microerror.Maskf(invalidConfigError, "config.K8sClient must not be empty")
-	}
+func New(config Config) (*Forwarder, error) {
 	if config.RestConfig == nil {
 		return nil, microerror.Maskf(invalidConfigError, "config.RestConfig must not be empty")
 	}
-	if config.PodName == "" {
-		return nil, microerror.Maskf(invalidConfigError, "config.PodName must not be empty")
+
+	k8sClient, err := rest.RESTClientFor(config.RestConfig)
```
Comment (kopiczko) [inline tunnel.go]: @fgimenez I think I found a cool way to fix both issues :)

File: tunnel.go (hunk 3 above)
Comment (kopiczko) [reply tunnel.go]: Added some godoc.

**Comment by fgimenez** [comment]
Thx, `rest.Interface` really makes a difference.

**Review by fgimenez — APPROVED** [review]

**Pull request merged**